---
Tom Curran

Problem Set \#4

MACS 30100

Monday February 5, 2018

---

2. **Classification problem**: KKN by hand and in Python (3 points). The table below provides a training data set containing six observations, three pre- dictors, and one qualitative response variable.

|Obs |$X_1$ |$X_2$ |$X_3$ |Y    |Dist from  $X_1$ |Dist from  $X_2$ |Dist from  $X_2$ |
|:--:|:----:|:----:|:----:|:---:|:---------------:|:---------------:|:---------------:|
|1   |0     |3     |0     |Red  |                 |                 |                 |
|2   |2     |0     |0     |Red  |                 |                 |                 |
|3   |0     |1     |3     |Red  |                 |                 |                 |
|4   |0     |1     |2     |Green|                 |                 |                 |
|5   |-1    |0     |1     |Green|                 |                 |                 |
|6   |1     |1     |1     |Red  |                 |                 |                 |
Suppose we wish to use this data set to make a prediction for Y when X1 = X2 = X3 = 0 using K-nearest neighbors.

(a) Compute the Euclidean distance between each observation and the test pointX1 =X2 =X3 =0.

$$
dist(p,q) = \sqrt{(q_1 - p_1)^2 +(q_2 - p_2)^2 +(q_3 - p_3)^2 }
$$

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
from statsmodels.iolib.summary2 import summary_col

/Users/thomascurran/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


In [119]:
df = pd.DataFrame({'obs': [1,2,3,4,5,6],
                  'x1':[0,2,0,0,-1,1],
                  'x2':[3,0,1,1,0,1],
                  'x3':[0,0,3,2,1,1],
                  'y':['Red','Red','Red','Green','Green','Red']})

df = df.set_index("obs")

df = df[['x1','x2','x3','y']]

In [120]:
df

x1  x2  x3      y
obs                   
1     0   3   0    Red
2     2   0   0    Red
3     0   1   3    Red
4     0   1   2  Green
5    -1   0   1  Green
6     1   1   1    Red

In [134]:
'''

distance = sqrt((x1 - 0)^2 +(x2 - 0)^2 +(x3 - 0)^2)

'''


def distance(q, p):
    q1,q2,q3 = q
    p1, p2, p3 = p
    
    dist1 =( q1 - p1 )** 2
    
    dist2 =( q2 - p2 )** 2
    
    dist3 =( q3 - p3 )** 2
    
    total_distance = np.sqrt(dist1 + dist2 + dist3)
    
    return(dist1, dist2, dist3, total_distance)

In [137]:
total_dist = []
dist_x1 = []
dist_x2 = []
dist_x3 = []

for i in range(0,len(df.index)):
    q_points = np.array(df.iloc[i][0:3])
    p_points = np.array([0,0,0]) 

    dist = distance(q_points,p_points)
    
    dist_x1.append(np.sqrt(dist[0]))
    dist_x2.append(np.sqrt(dist[1]))
    dist_x3.append(np.sqrt(dist[2]))
    
    total_dist.append(dist[3])
    
df['dist_from_x1'] = dist_x1
df['dist_from_x2'] = dist_x2
df['dist_from_x3'] = dist_x3
df['total_distance'] = total_dist

In [138]:
df

x1  x2  x3      y  dist_from_x1  dist_from_x2  dist_from_x3  \
obs                                                                
1     0   3   0    Red           0.0           3.0           0.0   
2     2   0   0    Red           2.0           0.0           0.0   
3     0   1   3    Red           0.0           1.0           3.0   
4     0   1   2  Green           0.0           1.0           2.0   
5    -1   0   1  Green           1.0           0.0           1.0   
6     1   1   1    Red           1.0           1.0           1.0   

     total_distance  
obs                  
1          3.000000  
2          2.000000  
3          3.162278  
4          2.236068  
5          1.414214  
6          1.732051

b) What is our KNN prediction with K = 1? Why?

For K = 1 and a predicted values of $X_i = (0,0,0)$, the closest neighbor for this value is $X_5 = 1.414214$. This is because according to KNN you want to categorize unknown values according to their closest known observations, which in this case if $X_5 = (-1,0,1)$. IN this case, since $K = 1$, we can predict that $X_i = (0,0,0)$ will be 'Green'

c)What is our KNN prediction with K = 3? Why?

When $K = 3$, and $X_i = (0,0,0)$ the closest neighbors are $X_2 = (4,0,0),\ X_5 = (-1,0,1)\ and\ X_6 =(1,1,1)$. Given these closest neighbors, $X_2,\ and\ X_6$ are 'red' while $X_5$ is green. Since there are 2 'red' and 1 'green' $X_i$ should be 'red'

d) If the Bayes (optimal) decision boundary in this problem is highly non- linear, then would we expect the best value for K to be large or small? Why?

The K value for a highly non-linear Bayes Decsion Boundary should be small. This is because a small K reduces smoothness of the boundary line that accounts for larger variance in the data and reduces bias. 

e)Use Python’s scikit-learn library to estimate the KNN classifier of the testpointX1 =X2 =X3 =1withK=2.

In [163]:
from sklearn import neighbors

x = df.as_matrix(columns = ['x1','x2','x3'])

y = np.array(df.y)

knn = neighbors.KNeighborsClassifier(n_neighbors = 2)

knn.fit(x,y)

x_j = np.array([1,1,1])

prediction = knn.predict([x_j])

neighbors = knn.kneighbors([x_j])

In [166]:
print("Predicted Y Value: ", prediction[0])
print("KNN Neigbors: ", neighbors)

Predicted Y Value:  Green
KNN Neigbors:  (array([[ 0.        ,  1.41421356]]), array([[5, 3]]))
